### PR TITLE
[SPARK-35044][SQL] `SET propertyKey` shall also lookup `sparkSession.sharedState.hadoopConf` to display the effective default hive/hadoop configs

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Currently, pure SQL users are short of ways to see the Hadoop configurations which may affect their jobs a lot, they are only able to get the Hadoop configs that exist in `SQLConf` while other defaults in `SharedState.hadoopConf` display wrongly and confusingly with `<undefined>`.

The pre-loaded ones from `core-site.xml, hive-site.xml` etc., will only stay in `sparkSession.sharedState.hadoopConf` or `sc._hadoopConfiguation` not `SQLConf`. Some of them that related the Hive Metastore connection(never change it spark runtime), e.g. `hive.metastore.uris`, are clearly global static and unchangeable but displayable I guess. Some of the ones that might be related to, for example, the output codec/compression, preset in Hadoop/hive config files like core-site.xml shall be still changeable from case to case, table to table, file to file, etc. It' meaningfully to show the defaults for users to change based on that.

In this PR, I propose to support get a Hadoop configuration by SET syntax, for example
```
SET mapreduce.map.output.compress.codec;
```

### Why are the changes needed?

better user experience for pure SQL users

### Does this PR introduce _any_ user-facing change?

yes, where retrieving a conf only existing in sessionState.hadoopConf, before is `undefined` and now you see it

### How was this patch tested?

new test

### CODE REVIEW

**Comment by hvanhovell** [comment]
I am not entirely sure that using SET syntax for reading a value is a good idea. People will expect that you can also modify them, which IMO should not be supported.

**Comment by yaooqinn** [comment]
> People will expect that you can also modify them, which IMO should not be supported.

We currently do need to support some Hive/Hadoop configurations at runtime, (for example, the STS), such as turning off/on the compression via `mapreduce.map.output.compress.codec`, which is modifiable but not able to see the default value. Returning the `<undefined>` causes ambiguity as well.

**Comment by hvanhovell** [comment]
I respectfully disagree. The problem is that you can have multiple tenants on the same cluster. People will become very unhappy when you allow users to change global configurations. If you want this done, you will need to introduce hadoop configurations at the session level.

**Comment by yaooqinn** [comment]
> If you want this done, you will need to introduce hadoop configurations at the session level.

The current behavior of the PR is equivalent to `sparkSession.sessionState.newHadoopConf().get(key)`, I guess this meets your point of ` the session level`. 

```scala
private[sql] object SessionState {
  def newHadoopConf(hadoopConf: Configuration, sqlConf: SQLConf): Configuration = {
    val newHadoopConf = new Configuration(hadoopConf)
    sqlConf.getAllConfs.foreach { case (k, v) => if (v ne null) newHadoopConf.set(k, v) }
    newHadoopConf
  }
}

```
Using the global `sharedState.hadoopConf` here is only to avoid the unneccary copy.

As I don't change the write-side, I believe it's also  at the session level.

**Comment by yaooqinn** [comment]
> People will become very unhappy when you allow users to change global configurations.
The changes made are not global. As a system administrator of the computing layer(e.g. me), personally, I am happy that the end-users can solve their problems case by case with our suggestions and dynamically verify and change the relative hive/hadoop configs at the session-level. On the contract, we will probably be unhappy when we need to change it globally at the server-side, which is very likely to affect other users' jobs and cause reboot of a resident service.

**Comment by yaooqinn** [comment]
also cc @cloud-fan @maropu @wangyum @HyukjinKwon and thanks very much

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala [hunk 1]
```diff
@@ -146,7 +146,9 @@ case class SetCommand(kv: Option[(String, Option[String])])
     // Queries a single property.
     case Some((key, None)) =>
       val runFunc = (sparkSession: SparkSession) => {
-        val value = sparkSession.conf.getOption(key).getOrElse("<undefined>")
+        val value = sparkSession.conf.getOption(key).getOrElse {
+          sparkSession.sharedState.hadoopConf.get(key, "<undefined>")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala:150]: Can we make the PR title clearer? It seems the SET command can already set/get hadoop configs in the SQLConf. The only problem is we don't display the default value correctly, which is from `sparkSession.sharedState.hadoopConf`.

BTW do we have a valid use case? e.g. a hadoop conf is not in `SQLConf` but in `sparkSession.sharedState.hadoopConf`.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala:150]: > Can we make the PR title clearer?

OK

> BTW do we have a valid use case? e.g. a hadoop conf is not in `SQLConf` but in `sparkSession.sharedState.hadoopConf`.

The pre-loaded ones from `core-site.xml, hive-site.xml` etc., will only stay in `sparkSession.sharedState.hadoopConf` or `sc. _hadoopConfiguation` not `SQLConf`. Some of them that related the Hive Metastore connection(never change it spark runtime), e.g. `hive.metastore.uris`, are clearly global static and unchangeable but displayable I guess. Some of the ones that might be related to, for example, the output codec/compression, preset in Hadoop/hive config files like `core-site.xml` shall bestill changeable from case to case, table to table, file to file, etc.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -1069,6 +1069,13 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     spark.sessionState.conf.clear()
   }
 
+  test("Support retrieve hadoop configuration via set command") {
+    val key = "hadoop.this.is.a.test.key"
+    val value = "2018-11-17 13:33:33.333"
+    spark.sharedState.hadoopConf.set(key, value)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1075]: Can we test with a valid use case? This doesn't seem to be valid.

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1075]: OK, I will add check for keys in `src/test/resources/hive-site.xml`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala:150]: let's add a few comments here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -1069,6 +1069,15 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     spark.sessionState.conf.clear()
   }
 
+  test("SPARK-35044: `SET propertyKey` shall lookup sharedState.hadoopConf to display the" +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1072]: nit: `SPARK-35044: SET command shall display default value for hadoop conf correctly`

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -1069,6 +1069,21 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+    // these keys are located at `src/test/resources/hive-site.xml`
+    checkAnswer(sql(s"SET $key"), Row(key, value))
+    checkAnswer(sql("SET hadoop.tmp.dir"), Row("hadoop.tmp.dir", "/tmp/hive_one"))
+
+    // this key does not exist
+    checkAnswer(sql(s"SET ${key}no"), Row(key + "no", "<undefined>"))
+    checkAnswer(sql("SET dfs.replication"), Row("dfs.replication", "<undefined>"))
+
+    // io.file.buffer.size has a default value from `SparkHadoopUtil.newConfiguration`
+    checkAnswer(sql("SET io.file.buffer.size"), Row("io.file.buffer.size", "65536"))
```
Comment (yaooqinn) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1084]: w/o the patch
```scala
spark-sql> SET io.file.buffer.size;
io.file.buffer.size	<undefined>
```

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Jenkins passed for many empty commits the GA also passed before rebasing master, the `fast forward` is not related

Thanks for reviewing and merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 5]
```diff
@@ -1069,6 +1069,21 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+  test("SPARK-35044: SET command shall display default value for hadoop conf correctly") {
+    val key = "hadoop.this.is.a.test.key"
+    val value = "2018-11-17 13:33:33.333"
+    // these keys are located at `src/test/resources/hive-site.xml`
+    checkAnswer(sql(s"SET $key"), Row(key, value))
+    checkAnswer(sql("SET hadoop.tmp.dir"), Row("hadoop.tmp.dir", "/tmp/hive_one"))
+
+    // these keys does not exist as default yet
+    checkAnswer(sql(s"SET ${key}no"), Row(key + "no", "<undefined>"))
+    checkAnswer(sql("SET dfs.replication"), Row("dfs.replication", "<undefined>"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1081]: This breaks the Hadoop 2.7 tests, seems the behavior is different between hadoop 2.7 and hadoop 3: https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test/job/spark-master-test-sbt-hadoop-2.7/2364/testReport/org.apache.spark.sql/SQLQuerySuite/SPARK_35044__SET_command_shall_display_default_value_for_hadoop_conf_correctly/

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1081]: I picked a different config which seems to work: `dfs.hosts`

@yaooqinn can you create a followup PR to try it?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:1081]: OK

**Pull request closed without merging**